File: array/array7.go
package array

import "fmt"

func plusOne(digits []int) []int {
	next := 0
	val := 0

	for i := len(digits) - 1; i >= 0; i-- {
		if i == len(digits)-1 {
			next = 1
		}
		val = digits[i] + next
		digits[i] = val % 10
		next = val / 10

	}

	if next > 0 {
		digits = append([]int{next}, digits...)
		fmt.Println(digits)
	}
	fmt.Println(digits)
	return digits
}

func plusOne2(digits []int) []int {
	for i := len(digits) - 1; i >= 0; i-- {
		if digits[i] < 9 {
			digits[i]++
			return digits
		}
		digits[i] = 0
	}
	digits = append([]int{1}, digits...)
	return digits
}

File: old/Array/array7.ts
// 코드 1
function plusOne(digits: number[]): number[] {
  const digit: bigint = BigInt(digits.join("")) + BigInt(1);
  return digit
    .toString()
    .split("")
    .map((c) => +c);

  // return (BigInt(digits.join('')) + BigInt(1)).toString().split('').map((c) => +c);
}

// 코드 2
function plusOne(digits: number[]): number[] {
  ++digits[digits.length - 1];
  let carry: number = 0;

  for (let i: number = digits.length - 1; i >= 0; i--) {
    digits[i] += carry;
    carry = 0;

    if (digits[i] > 9) {
      digits[i] -= 10;
      carry = 1;
    }
  }

  if (carry) digits.unshift(1);
  return digits;
}
